### PR TITLE
[Bug] Comparison between two PeriodIndexes doesn't validate length (GH #23078)

### LINKED ISSUE #23078: comparison between two PeriodIndexes doesn't validate length
```python
In [5]: idx = pd.period_range('2017', periods=12, freq="A-DEC")

In [6]: idx <= idx[[0]]
Out[6]:
array([ True, False, False, False, False, False, False, False, False,
       False, False, False])

```

This should raise.

Both `DatetimeIndex` and `TimedeltaIndex` do raise

```python
In [8]: idx <= idx[[0]]
```

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-8a2a887e6efd> in <module>
----> 1 idx <= idx[[0]]

~/sandbox/pandas/pandas/core/indexes/datetimes.py in wrapper(self, other)
     90
     91     def wrapper(self, other):
---> 92         result = getattr(DatetimeArrayMixin, opname)(self, other)
     93         if is_bool_dtype(result):
     94             return result

~/sandbox/pandas/pandas/core/arrays/datetimes.py in wrapper(self, other)
    133             else:
    134                 self._assert_tzawareness_compat(other)
--> 135                 result = meth(self, np.asarray(other))
    136
    137             result = com.values_from_object(result)

~/sandbox/pandas/pandas/core/arrays/datetimelike.py in cmp_method(self, other)
     51         if isinstance(other, (np.ndarray, ABCIndexClass, ABCSeries)):
     52             if other.ndim > 0 and len(self) != len(other):
---> 53                 raise ValueError('Lengths must match to compare')
     54
     55         if needs_i8_conversion(self) and needs_i8_conversion(other):

ValueError: Lengths must match to compare

```

### PR DESCRIPTION
**Pull request by makbigc**
- [x] closes #23078
- [x] tests added
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/period/test_period.py [hunk 1]
```diff
@@ -561,6 +561,12 @@ def test_insert(self):
             result = period_range('2017Q1', periods=4, freq='Q').insert(1, na)
             tm.assert_index_equal(result, expected)
 
+    def test_comp_op(self):
```
Comment (TomAugspurger) [inline pandas/tests/indexes/period/test_period.py]: Slight nitpick: this test name isn't that descriptive of what it's testing. I'd prefer something like `test_comparing_different_lengths_raises`.

**Review by TomAugspurger — APPROVED** [review]

File: pandas/core/arrays/period.py [hunk 2]
```diff
@@ -73,6 +73,9 @@ def wrapper(self, other):
                 msg = DIFFERENT_FREQ_INDEX.format(self.freqstr, other.freqstr)
                 raise IncompatibleFrequency(msg)
 
+            if other.ndim > 0 and len(self) != len(other):
```
Comment (jreback) [inline pandas/core/arrays/period.py]: this should fail if ndim of other != 1

we don't generally check this for other arrays (though we should). So I would maybe add this check on the base array class and just call it here, ``_validate_equal(self, other)`` or something

**Review by jreback — CHANGES_REQUESTED** [review]
let's make this a general check in the base class instead

File: pandas/core/arrays/period.py (hunk 2 above)
Comment (jreback) [reply pandas/core/arrays/period.py]: yes, I think we have an implicit guarantee on 1-D ness now (or at the very least the pandas extension types do).

Comment (jbrockmendel) [reply pandas/core/arrays/period.py]: Related #23853

**Comment by jreback** [comment]
@makbigc can you merge master and update

**Comment by jreback** [comment]
this overlaps with #24397 @jbrockmendel can you comment here

File: pandas/tests/extension/decimal/test_decimal.py [hunk 3]
```diff
@@ -299,6 +299,12 @@ def _check_divmod_op(self, s, op, other, exc=NotImplementedError):
     def test_error(self):
         pass
 
+    def test_arith_diff_lengths(self):
```
Comment (jbrockmendel) [inline pandas/tests/extension/decimal/test_decimal.py]: are these tests `pass`ed because they would fail?  if so, mark with `@pytest.mark.xfail`.

File: pandas/tests/extension/base/ops.py [hunk 4]
```diff
@@ -119,6 +119,12 @@ def test_direct_arith_with_series_returns_not_implemented(self, data):
                 "{} does not implement add".format(data.__class__.__name__)
             )
 
+    def test_arith_diff_lengths(self, data, all_arithmetic_operators):
+        op = self.get_op_from_name(all_arithmetic_operators)
+        other = data[:3]
```
Comment (jbrockmendel) [inline pandas/tests/extension/base/ops.py]: This test actually wouldn't catch the behavior in #23078, which is caused specifically by comparisons of PeriodIndex with length-1 objects.

Comment (makbigc) [reply pandas/tests/extension/base/ops.py]: The issue in #23078 was treated in the function test_comp_op (pandas/tests/indexes/period/test_period.py).

File: doc/source/whatsnew/v0.24.0.rst [hunk 5]
```diff
@@ -1319,6 +1319,7 @@ Datetimelike
 - Bug in :class:`DatetimeIndex` and :class:`TimedeltaIndex` where indexing with ``Ellipsis`` would incorrectly lose the index's ``freq`` attribute (:issue:`21282`)
 - Clarified error message produced when passing an incorrect ``freq`` argument to :class:`DatetimeIndex` with ``NaT`` as the first entry in the passed data (:issue:`11587`)
 - Bug in :func:`to_datetime` where ``box`` and ``utc`` arguments were ignored when passing a :class:`DataFrame` or ``dict`` of unit mappings (:issue:`23760`)
+- Bug in :class:`PeriodIndex` when comparing indexes of different lengths, ValueError is not raised (:issue:`23078`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: use double backticks around ValueError

File: pandas/tests/extension/base/ops.py [hunk 6]
```diff
@@ -119,6 +119,12 @@ def test_direct_arith_with_series_returns_not_implemented(self, data):
...
+        with pytest.raises(ValueError):
```
Comment (jreback) [inline pandas/tests/extension/base/ops.py]: can you match the message as well

File: pandas/tests/extension/base/ops.py [hunk 7]
```diff
@@ -164,3 +170,9 @@ def test_direct_arith_with_series_returns_not_implemented(self, data):
             raise pytest.skip(
                 "{} does not implement __eq__".format(data.__class__.__name__)
             )
+
+    def test_compare_diff_lengths(self, data, all_compare_operators):
+        op = self.get_op_from_name(all_compare_operators)
+        other = data[:3]
+        with pytest.raises(ValueError):
```
Comment (jreback) [inline pandas/tests/extension/base/ops.py]: match the message

File: pandas/tests/extension/decimal/test_decimal.py [hunk 8]
```diff
@@ -324,6 +331,12 @@ def test_compare_array(self, data, all_compare_operators):
                                    for i in alter]
         self._compare_other(s, data, op_name, other)
 
+    # TODO:
```
Comment (jreback) [inline pandas/tests/extension/decimal/test_decimal.py]: why are these xfail? make the change as needed in the decimal code

File: pandas/tests/extension/json/test_json.py [hunk 9]
```diff
@@ -291,9 +291,13 @@ def _check_divmod_op(self, s, op, other, exc=NotImplementedError):
             s, op, other, exc=TypeError
         )
 
+    def test_arith_diff_lengths(self):
```
Comment (jreback) [inline pandas/tests/extension/json/test_json.py]: why are these added?

**Review by jreback — CHANGES_REQUESTED** [review]
why are you adding tests to each of the EA subclasses? the base test should work for each of these.

**Comment by makbigc** [comment]
Should I close this PR? Because #23078 was already resolved by #24397. The length check is done in the wrapper of _period_array_cmp. This is not need to add _validate_shape additionally.

The generic base test cannot handle all six EA subclasses:
- Integer array: deprecationwarning raised by numpy
- category: only __eq__ and __ne__ comparison are implemented
- json: no json comparison

The raise of ValueError for DecimalArray could be implemented in a new PR. 
@jreback @TomAugspurger  Thank you for your review.
